File: peaky_finders/src/peaky_finders/serve/land_query.py
"""Point-in-polygon queries against preset land layers."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime
from decimal import Decimal
from pathlib import Path
from typing import Any

from shapely.geometry import Point
from shapely.strtree import STRtree

from peaky_finders.core.preset import LandLayerEntry, Preset, load_preset
from peaky_finders.serve.land_import import read_land_layer_gdf, resolve_land_source_path


@dataclass(frozen=True)
class LandPointHit:
    """One land polygon containing the query point."""

    properties: dict[str, Any]
    area: float


@dataclass
class LandLayerSpatialIndex:
    """Spatial index for repeated point queries against one land layer."""

    tree: STRtree
    geoms: list
    hits: list[LandPointHit]


def _json_safe_value(value: Any) -> Any:
    if value is None:
        return None
    try:
        import pandas as pd

        if pd.isna(value):
            return None
    except (TypeError, ValueError):
        pass
    if isinstance(value, bool):
        return value
    if isinstance(value, int) and not isinstance(value, bool):
        return int(value)
    if isinstance(value, float):
        return float(value)
    if isinstance(value, str):
        s = value.strip()
        return s or None
    if isinstance(value, (datetime, date)):
        return value.isoformat()
    if isinstance(value, Decimal):
        return float(value)
    return str(value)


def _row_properties(row: Any, columns: list[str]) -> dict[str, Any]:
    props: dict[str, Any] = {}
    for col in columns:
        if col == "geometry":
            continue
        safe = _json_safe_value(row[col])
        if safe is not None:
            props[col] = safe
    return props


def resolve_land_layer_entry(
    preset: Preset,
    source_id: str,
    layer_name: str | None = None,
) -> LandLayerEntry:
    source = preset.land.sources.get(source_id)
    if source is None:
        raise ValueError(f"unknown land source: {source_id}")
    if layer_name is None:
        if len(source.layers) != 1:
            raise ValueError(
                f"land source {source_id!r} has {len(source.layers)} layers; pass layer_name"
            )
        return source.layers[0]
    for layer in source.layers:
        if layer.name == layer_name:
            return layer
    names = [layer.name for layer in source.layers]
    raise ValueError(f"layer {layer_name!r} not found on {source_id!r}; have {names}")


def load_land_layer_index(
    preset_path: Path,
    source_id: str,
    layer_name: str | None = None,
    *,
    entry: LandLayerEntry | None = None,
) -> LandLayerSpatialIndex:
    """Build a spatial index for one preset land layer."""
    preset_path = Path(preset_path).expanduser().resolve()
    preset = load_preset(preset_path)
    project_dir = preset_path.parent
    source = preset.land.sources.get(source_id)
    if source is None:
        raise ValueError(f"unknown land source: {source_id}")

    layer_entry = entry or resolve_land_layer_entry(preset, source_id, layer_name)
    data_path = resolve_land_source_path(project_dir, source.path)
    gdf = read_land_layer_gdf(data_path, layer_entry)
    if gdf.empty:
        return LandLayerSpatialIndex(tree=STRtree([]), geoms=[], hits=[])

    columns = [str(c) for c in gdf.columns]
    geoms: list = []
    hits: list[LandPointHit] = []
    for _, row in gdf.iterrows():
        geom = row.geometry
        if geom is None or geom.is_empty:
            continue
        if not geom.is_valid:
            geom = geom.buffer(0)
        geoms.append(geom)
        hits.append(
            LandPointHit(
                properties=_row_properties(row, columns),
                area=float(geom.area),
            )
        )

    return LandLayerSpatialIndex(tree=STRtree(geoms), geoms=geoms, hits=hits)


def query_land_layer_index(
    index: LandLayerSpatialIndex,
    lat: float,
    lon: float,
    *,
    smallest_only: bool = False,
) -> list[LandPointHit]:
    """Return land features containing ``(lat, lon)``."""
    if not index.geoms:
        return []

    pt = Point(float(lon), float(lat))
    matched: list[LandPointHit] = []
    for idx in index.tree.query(pt):
        if index.geoms[idx].contains(pt):
            matched.append(index.hits[idx])

    if not matched:
        return []
    if smallest_only and len(matched) > 1:
        return [min(matched, key=lambda hit: hit.area)]
    return matched


def point_hits(
    preset_path: Path,
    lat: float,
    lon: float,
    source_id: str,
    layer_name: str | None = None,
    *,
    entry: LandLayerEntry | None = None,
    smallest_only: bool = False,
) -> list[dict[str, Any]]:
    """Point query; returns attribute dicts for each matching polygon."""
    index = load_land_layer_index(
        preset_path,
        source_id,
        layer_name,
        entry=entry,
    )
    return [hit.properties for hit in query_land_layer_index(index, lat, lon, smallest_only=smallest_only)]

File: crates/peaky-geo/src/land_query.rs
//! Point-in-polygon queries against preset land layers (GeoJSON).

use std::path::Path;

use anyhow::{bail, Context, Result};
use geo::{Area, Contains, Geometry, HasDimensions, Point};
use geo::algorithm::bounding_rect::BoundingRect;
use geojson::{Feature, GeoJson, Value as GeoJsonValue};
use peaky_preset::{load_preset, LandLayerEntry, Preset};
use rstar::{RTree, RTreeObject, AABB};

use crate::land_filter::{json_safe_properties, matches_land_attribute_filters};
use crate::land_path::{is_land_geojson_path, resolve_land_layer_geojson_path};

#[derive(Debug, Clone)]
pub struct LandPointHit {
    pub properties: serde_json::Map<String, serde_json::Value>,
    pub area: f64,
}

struct IndexedPolygon {
    envelope: AABB<[f64; 2]>,
    index: usize,
}

impl RTreeObject for IndexedPolygon {
    type Envelope = AABB<[f64; 2]>;

    fn envelope(&self) -> Self::Envelope {
        self.envelope
    }
}

pub struct LandLayerSpatialIndex {
    tree: RTree<IndexedPolygon>,
    geoms: Vec<Geometry<f64>>,
    hits: Vec<LandPointHit>,
}

pub fn resolve_land_layer_entry(
    preset: &Preset,
    source_id: &str,
    layer_name: Option<&str>,
) -> Result<LandLayerEntry> {
    let source = preset
        .land
        .sources
        .get(source_id)
        .with_context(|| format!("unknown land source: {source_id}"))?;
    if let Some(name) = layer_name {
        for layer in &source.layers {
            if layer.name == name {
                return Ok(layer.clone());
            }
        }
        let names: Vec<_> = source.layers.iter().map(|l| l.name.as_str()).collect();
        bail!("layer {name:?} not found on {source_id:?}; have {names:?}");
    }
    if source.layers.len() != 1 {
        bail!(
            "land source {source_id:?} has {} layers; pass layer_name",
            source.layers.len()
        );
    }
    Ok(source.layers[0].clone())
}

fn geometry_from_geojson(value: &GeoJsonValue) -> Result<Geometry<f64>> {
    let geom: geo::Geometry<f64> = value
        .clone()
        .try_into()
        .context("convert GeoJSON geometry")?;
    Ok(geom)
}

fn polygon_envelope(geom: &Geometry<f64>) -> Option<AABB<[f64; 2]>> {
    let rect = geom.bounding_rect()?;
    Some(AABB::from_corners(
        [rect.min().x, rect.min().y],
        [rect.max().x, rect.max().y],
    ))
}

pub fn read_land_layer_features(
    data_path: &Path,
    entry: &LandLayerEntry,
) -> Result<Vec<(Geometry<f64>, LandPointHit)>> {
    if !is_land_geojson_path(data_path) {
        bail!("GeoJSON-only land import: {}", data_path.display());
    }
    let text = std::fs::read_to_string(data_path)
        .with_context(|| format!("read {}", data_path.display()))?;
    let geojson: GeoJson = text.parse().context("parse GeoJSON")?;
    let features = match geojson {
        GeoJson::FeatureCollection(fc) => fc.features,
        GeoJson::Feature(f) => vec![f],
        _ => bail!("expected GeoJSON FeatureCollection"),
    };

    let mut rows = Vec::new();
    for feature in features {
        let Feature {
            geometry,
            properties,
            ..
        } = feature;
        let Some(geometry) = geometry else {
            continue;
        };
        let props_map = properties.unwrap_or_default();
        if !matches_land_attribute_filters(&props_map, &entry.include, &entry.exclude) {
            continue;
        }
        let geom = geometry_from_geojson(&geometry.value)?;
        if geom.is_empty() {
            continue;
        }
        let area = geom.unsigned_area();
        let safe = json_safe_properties(&props_map);
        rows.push((
            geom,
            LandPointHit {
                properties: safe,
                area,
            },
        ));
    }
    Ok(rows)
}

pub fn load_land_layer_index(
    preset_path: &Path,
    source_id: &str,
    layer_name: Option<&str>,
    entry: Option<&LandLayerEntry>,
) -> Result<LandLayerSpatialIndex> {
    let preset = load_preset(preset_path)?;
    let project_dir = preset_path
        .parent()
        .context("preset path must have a parent directory")?;
    let source = preset
        .land
        .sources
        .get(source_id)
        .with_context(|| format!("unknown land source: {source_id}"))?;
    let layer_entry = match entry {
        Some(e) => e.clone(),
        None => resolve_land_layer_entry(&preset, source_id, layer_name)?,
    };
    let data_path = resolve_land_layer_geojson_path(
        project_dir,
        source_id,
        &layer_entry.name,
        &source.path,
    )?;
    let rows = read_land_layer_features(&data_path, &layer_entry)?;

    let mut geoms = Vec::with_capacity(rows.len());
    let mut hits = Vec::with_capacity(rows.len());
    let mut indexed = Vec::with_capacity(rows.len());
    for (geom, hit) in rows {
        if let Some(envelope) = polygon_envelope(&geom) {
            let index = geoms.len();
            indexed.push(IndexedPolygon { envelope, index });
            geoms.push(geom);
            hits.push(hit);
        }
    }

    let tree = RTree::bulk_load(indexed);
    Ok(LandLayerSpatialIndex { tree, geoms, hits })
}

pub fn query_land_layer_index(
    index: &LandLayerSpatialIndex,
    lat: f64,
    lon: f64,
    smallest_only: bool,
) -> Vec<LandPointHit> {
    if index.geoms.is_empty() {
        return Vec::new();
    }

    let pt = Point::new(lon, lat);
    let mut matched = Vec::new();
    for item in index.tree.locate_in_envelope_intersecting(&AABB::from_point([lon, lat])) {
        let geom = &index.geoms[item.index];
        if geom.contains(&pt) {
            matched.push(index.hits[item.index].clone());
        }
    }

    if matched.is_empty() {
        return matched;
    }
    if smallest_only && matched.len() > 1 {
        matched.sort_by(|a, b| {
            a.area
                .partial_cmp(&b.area)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        vec![matched.remove(0)]
    } else {
        matched
    }
}

pub fn point_hits(
    preset_path: &Path,
    lat: f64,
    lon: f64,
    source_id: &str,
    layer_name: Option<&str>,
    entry: Option<&LandLayerEntry>,
    smallest_only: bool,
) -> Result<Vec<serde_json::Map<String, serde_json::Value>>> {
    let index = load_land_layer_index(preset_path, source_id, layer_name, entry)?;
    Ok(query_land_layer_index(&index, lat, lon, smallest_only)
        .into_iter()
        .map(|hit| hit.properties)
        .collect())
}

